In [1]:
import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
cars = pd.read_csv('CarPrice_Assignment.csv', encoding='utf-8', delimiter=',')
cars

car_ID  symboling                   CarName fueltype aspiration  \
0         1          3        alfa-romero giulia      gas        std   
1         2          3       alfa-romero stelvio      gas        std   
2         3          1  alfa-romero Quadrifoglio      gas        std   
3         4          2               audi 100 ls      gas        std   
4         5          2                audi 100ls      gas        std   
..      ...        ...                       ...      ...        ...   
200     201         -1           volvo 145e (sw)      gas        std   
201     202         -1               volvo 144ea      gas      turbo   
202     203         -1               volvo 244dl      gas        std   
203     204         -1                 volvo 246   diesel      turbo   
204     205         -1               volvo 264gl      gas      turbo   

    doornumber      carbody drivewheel enginelocation  wheelbase  ...  \
0          two  convertible        rwd          front       88.6  ...   
1          two  convertible        rwd          front       88.6  ...   
2          two    hatchback        rwd          front       94.5  ...   
3         four        sedan        fwd          front       99.8  ...   
4         four        sedan        4wd          front       99.4  ...   
..         ...          ...        ...            ...        ...  ...   
200       four        sedan        rwd          front      109.1  ...   
201       four        sedan        rwd          front      109.1  ...   
202       four        sedan        rwd          front      109.1  ...   
203       four        sedan        rwd          front      109.1  ...   
204       four        sedan        rwd          front      109.1  ...   

     enginesize  fuelsystem  boreratio  stroke compressionratio horsepower  \
0           130        mpfi       3.47    2.68              9.0        111   
1           130        mpfi       3.47    2.68              9.0        111   
2           152        mpfi       2.68    3.47              9.0        154   
3           109        mpfi       3.19    3.40             10.0        102   
4           136        mpfi       3.19    3.40              8.0        115   
..          ...         ...        ...     ...              ...        ...   
200         141        mpfi       3.78    3.15              9.5        114   
201         141        mpfi       3.78    3.15              8.7        160   
202         173        mpfi       3.58    2.87              8.8        134   
203         145         idi       3.01    3.40             23.0        106   
204         141        mpfi       3.78    3.15              9.5        114   

     peakrpm citympg  highwaympg    price  
0       5000      21          27  13495.0  
1       5000      21          27  16500.0  
2       5000      19          26  16500.0  
3       5500      24          30  13950.0  
4       5500      18          22  17450.0  
..       ...     ...         ...      ...  
200     5400      23          28  16845.0  
201     5300      19          25  19045.0  
202     5500      18          23  21485.0  
203     4800      26          27  22470.0  
204     5400      19          25  22625.0  

[205 rows x 26 columns]

the above data is having some cars that have same name but written in differnt way is also considred as a new category 
EX : audi 100 ls and audi 100ls	......same name but space diff.
    also applied for capiatal and small cases


In [3]:
cars.shape # Rows * Columns

(205, 26)

In [4]:
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   car_ID            205 non-null    int64  
 1   symboling         205 non-null    int64  
 2   CarName           205 non-null    object 
 3   fueltype          205 non-null    object 
 4   aspiration        205 non-null    object 
 5   doornumber        205 non-null    object 
 6   carbody           205 non-null    object 
 7   drivewheel        205 non-null    object 
 8   enginelocation    205 non-null    object 
 9   wheelbase         205 non-null    float64
 10  carlength         205 non-null    float64
 11  carwidth          205 non-null    float64
 12  carheight         205 non-null    float64
 13  curbweight        205 non-null    int64  
 14  enginetype        205 non-null    object 
 15  cylindernumber    205 non-null    object 
 16  enginesize        205 non-null    int64  
 1

PREPROCESSING


if we split 1st word or a word before space we will have Brand name so we here we are performing this

In [5]:
cars['CarName'].apply(lambda x: x.split(' ')[0]) # will take all carnames

0      alfa-romero
1      alfa-romero
2      alfa-romero
3             audi
4             audi
          ...     
200          volvo
201          volvo
202          volvo
203          volvo
204          volvo
Name: CarName, Length: 205, dtype: object

In [6]:
CompanyNames= cars['CarName'].apply(lambda x: x.split(' ')[0]) # apply function applies a function to each element in the Series called CarName
cars.insert(3, 'CompanyName', CompanyNames) # insert the new column CompanyName at index 3
cars.drop('CarName', axis=1, inplace=True) # drop the old CarName column
cars.head()

car_ID  symboling  CompanyName fueltype aspiration doornumber      carbody  \
0       1          3  alfa-romero      gas        std        two  convertible   
1       2          3  alfa-romero      gas        std        two  convertible   
2       3          1  alfa-romero      gas        std        two    hatchback   
3       4          2         audi      gas        std       four        sedan   
4       5          2         audi      gas        std       four        sedan   

  drivewheel enginelocation  wheelbase  ...  enginesize  fuelsystem  \
0        rwd          front       88.6  ...         130        mpfi   
1        rwd          front       88.6  ...         130        mpfi   
2        rwd          front       94.5  ...         152        mpfi   
3        fwd          front       99.8  ...         109        mpfi   
4        4wd          front       99.4  ...         136        mpfi   

   boreratio  stroke compressionratio horsepower  peakrpm citympg  highwaympg  \
0       3.47    2.68              9.0        111     5000      21          27   
1       3.47    2.68              9.0        111     5000      21          27   
2       2.68    3.47              9.0        154     5000      19          26   
3       3.19    3.40             10.0        102     5500      24          30   
4       3.19    3.40              8.0        115     5500      18          22   

     price  
0  13495.0  
1  16500.0  
2  16500.0  
3  13950.0  
4  17450.0  

[5 rows x 26 columns]

In [7]:
cars.CompanyName.unique() # unique values in the CompanyName column

array(['alfa-romero', 'audi', 'bmw', 'chevrolet', 'dodge', 'honda',
       'isuzu', 'jaguar', 'maxda', 'mazda', 'buick', 'mercury',
       'mitsubishi', 'Nissan', 'nissan', 'peugeot', 'plymouth', 'porsche',
       'porcshce', 'renault', 'saab', 'subaru', 'toyota', 'toyouta',
       'vokswagen', 'volkswagen', 'vw', 'volvo'], dtype=object)

Fixing invalid values

In [8]:
# we have to correct some company names
cars.CompanyName = cars.CompanyName.str.lower() # convert all company names to lower case
def replace_name(a,b):
    cars.CompanyName.replace(a, b, inplace=True) # replace the old name with the new name

replace_name('maxda', 'mazda')
replace_name('porcshce', 'porsche')
replace_name('toyouta', 'toyota')
replace_name('vokswagen', 'volkswagen')
replace_name('vw', 'volkswagen')
replace_name('chevroelt', 'chevrolet')

cars.CompanyName.unique() # check the unique values again


array(['alfa-romero', 'audi', 'bmw', 'chevrolet', 'dodge', 'honda',
       'isuzu', 'jaguar', 'mazda', 'buick', 'mercury', 'mitsubishi',
       'nissan', 'peugeot', 'plymouth', 'porsche', 'renault', 'saab',
       'subaru', 'toyota', 'volkswagen', 'volvo'], dtype=object)

In [9]:
cars.columns # check the columns in the dataframe

Index(['car_ID', 'symboling', 'CompanyName', 'fueltype', 'aspiration',
       'doornumber', 'carbody', 'drivewheel', 'enginelocation', 'wheelbase',
       'carlength', 'carwidth', 'carheight', 'curbweight', 'enginetype',
       'cylindernumber', 'enginesize', 'fuelsystem', 'boreratio', 'stroke',
       'compressionratio', 'horsepower', 'peakrpm', 'citympg', 'highwaympg',
       'price'],
      dtype='object')

FEATURE ENGINEERING

In [10]:
# Derive new column - fuel economy
cars['fuel_economy']= (0.55 * cars['citympg']) + (0.45 * cars['highwaympg']) # giving more weigtage to city milage is 55 % and for highway milage is 45 %

In [11]:
cars['price'] = cars['price'].astype(int) # convert price column to int type
temp = cars.copy() # create a copy of the dataframe

In [12]:
temp.groupby(['CompanyName'])['price'].mean()

CompanyName
alfa-romero    15498.333333
audi           17859.142857
bmw            26118.750000
buick          33647.000000
chevrolet       6007.000000
dodge           7875.444444
honda           8184.692308
isuzu           8916.250000
jaguar         34600.000000
mazda          10652.882353
mercury        16503.000000
mitsubishi      9239.769231
nissan         10415.666667
peugeot        15489.090909
plymouth        7963.428571
porsche        31400.400000
renault         9595.000000
saab           15223.333333
subaru          8541.250000
toyota          9885.812500
volkswagen     10077.500000
volvo          18063.181818
Name: price, dtype: float64

In [13]:
table = temp.groupby(['CompanyName'])['price'].mean() # group by CompanyName and calculate the mean price
temp = temp.merge(table.reset_index(),how= 'left',on= 'CompanyName') # merge the mean price with the original dataframe using left join

In [14]:
# have to derive new column - cartype (Budget , Medium, Highrange)
bins =[0,10000,20000,40000] # it is 0- 9999, 10000-19999, 20000-39999
cars_bin = ['Budget', 'Medium', 'Highend'] # define the bins for the price rang
cars['carsrange'] = pd.cut(temp['price_y'],bins,right=False, labels=cars_bin) # create bins for the price range
cars.head() # check the new column

car_ID  symboling  CompanyName fueltype aspiration doornumber      carbody  \
0       1          3  alfa-romero      gas        std        two  convertible   
1       2          3  alfa-romero      gas        std        two  convertible   
2       3          1  alfa-romero      gas        std        two    hatchback   
3       4          2         audi      gas        std       four        sedan   
4       5          2         audi      gas        std       four        sedan   

  drivewheel enginelocation  wheelbase  ...  boreratio  stroke  \
0        rwd          front       88.6  ...       3.47    2.68   
1        rwd          front       88.6  ...       3.47    2.68   
2        rwd          front       94.5  ...       2.68    3.47   
3        fwd          front       99.8  ...       3.19    3.40   
4        4wd          front       99.4  ...       3.19    3.40   

   compressionratio  horsepower peakrpm citympg  highwaympg  price  \
0               9.0         111    5000      21          27  13495   
1               9.0         111    5000      21          27  16500   
2               9.0         154    5000      19          26  16500   
3              10.0         102    5500      24          30  13950   
4               8.0         115    5500      18          22  17450   

   fuel_economy  carsrange  
0         23.70     Medium  
1         23.70     Medium  
2         22.15     Medium  
3         26.70     Medium  
4         19.80     Medium  

[5 rows x 28 columns]

FIND HIGH CORRELATIONS BETWEEN FEATURES (in linear regression all your Xs should be independent so we have to find the correlation and need to eliminate it)

In [15]:
cor = cars.select_dtypes(include=[np.number]).corr() # calculate the correlation matrix for numerical columns
cor

car_ID  symboling  wheelbase  carlength  carwidth  \
car_ID            1.000000  -0.151621   0.129729   0.170636  0.052387   
symboling        -0.151621   1.000000  -0.531954  -0.357612 -0.232919   
wheelbase         0.129729  -0.531954   1.000000   0.874587  0.795144   
carlength         0.170636  -0.357612   0.874587   1.000000  0.841118   
carwidth          0.052387  -0.232919   0.795144   0.841118  1.000000   
carheight         0.255960  -0.541038   0.589435   0.491029  0.279210   
curbweight        0.071962  -0.227691   0.776386   0.877728  0.867032   
enginesize       -0.033930  -0.105790   0.569329   0.683360  0.735433   
boreratio         0.260064  -0.130051   0.488750   0.606454  0.559150   
stroke           -0.160824  -0.008735   0.160959   0.129533  0.182942   
compressionratio  0.150276  -0.178515   0.249786   0.158414  0.181129   
horsepower       -0.015006   0.070873   0.353294   0.552623  0.640732   
peakrpm          -0.203789   0.273606  -0.360469  -0.287242 -0.220012   
citympg           0.015940  -0.035823  -0.470414  -0.670909 -0.642704   
highwaympg        0.011255   0.034606  -0.544082  -0.704662 -0.677218   
price            -0.109093  -0.079978   0.577816   0.682921  0.759325   
fuel_economy      0.013872  -0.003257  -0.508136  -0.691472 -0.663419   

                  carheight  curbweight  enginesize  boreratio    stroke  \
car_ID             0.255960    0.071962   -0.033930   0.260064 -0.160824   
symboling         -0.541038   -0.227691   -0.105790  -0.130051 -0.008735   
wheelbase          0.589435    0.776386    0.569329   0.488750  0.160959   
carlength          0.491029    0.877728    0.683360   0.606454  0.129533   
carwidth           0.279210    0.867032    0.735433   0.559150  0.182942   
carheight          1.000000    0.295572    0.067149   0.171071 -0.055307   
curbweight         0.295572    1.000000    0.850594   0.648480  0.168790   
enginesize         0.067149    0.850594    1.000000   0.583774  0.203129   
boreratio          0.171071    0.648480    0.583774   1.000000 -0.055909   
stroke            -0.055307    0.168790    0.203129  -0.055909  1.000000   
compressionratio   0.261214    0.151362    0.028971   0.005197  0.186110   
horsepower        -0.108802    0.750739    0.809769   0.573677  0.080940   
peakrpm           -0.320411   -0.266243   -0.244660  -0.254976 -0.067964   
citympg           -0.048640   -0.757414   -0.653658  -0.584532 -0.042145   
highwaympg        -0.107358   -0.797465   -0.677470  -0.587012 -0.043931   
price              0.119337    0.835305    0.874145   0.553174  0.079443   
fuel_economy      -0.076356   -0.781535   -0.669464  -0.589898 -0.043281   

                  compressionratio  horsepower   peakrpm   citympg  \
car_ID                    0.150276   -0.015006 -0.203789  0.015940   
symboling                -0.178515    0.070873  0.273606 -0.035823   
wheelbase                 0.249786    0.353294 -0.360469 -0.470414   
carlength                 0.158414    0.552623 -0.287242 -0.670909   
carwidth                  0.181129    0.640732 -0.220012 -0.642704   
carheight                 0.261214   -0.108802 -0.320411 -0.048640   
curbweight                0.151362    0.750739 -0.266243 -0.757414   
enginesize                0.028971    0.809769 -0.244660 -0.653658   
boreratio                 0.005197    0.573677 -0.254976 -0.584532   
stroke                    0.186110    0.080940 -0.067964 -0.042145   
compressionratio          1.000000   -0.204326 -0.435741  0.324701   
horsepower               -0.204326    1.000000  0.131073 -0.801456   
peakrpm                  -0.435741    0.131073  1.000000 -0.113544   
citympg                   0.324701   -0.801456 -0.113544  1.000000   
highwaympg                0.265201   -0.770544 -0.054275  0.971337   
price                     0.067984    0.808138 -0.085268 -0.685752   
fuel_economy              0.299310   -0.792822 -0.086739  0.993844   

                  highwaympg     price  fuel_economy  
car_ID              

whenever we have to understand the correlation we use heat map

In [23]:
# converting categorical columns to numerical columns
def dummies(x,df):
    temp = pd.get_dummies(df[x],drop_first=True, dtype=float) # create dummy variables for the categorical column, drop_first=True will drop the first category to avoid dummy variable trap
    df = pd.concat([df,temp],axis=1) # concatenate the dummy variables with the original dataframe
    df.drop([x], axis=1, inplace=True) # drop the original categorical column
    return df

In [24]:
# Applying the dummies function to the cars_lr
cars_lr = cars
cars_lr = dummies('fueltype', cars_lr)
cars_lr = dummies('aspiration', cars_lr)
cars_lr = dummies('carbody', cars_lr)
cars_lr = dummies('drivewheel', cars_lr)
cars_lr = dummies('enginetype', cars_lr)
cars_lr = dummies('cylindernumber', cars_lr)
cars_lr = dummies('carsrange', cars_lr)
cars_lr = dummies('CompanyName', cars_lr) # convert the CompanyName column to dummy variables
cars_lr = dummies('doornumber', cars_lr)
cars_lr = dummies('enginelocation', cars_lr)
cars_lr = dummies('fuelsystem', cars_lr) 


In [25]:
cars_lr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 69 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   car_ID            205 non-null    int64  
 1   symboling         205 non-null    int64  
 2   wheelbase         205 non-null    float64
 3   carlength         205 non-null    float64
 4   carwidth          205 non-null    float64
 5   carheight         205 non-null    float64
 6   curbweight        205 non-null    int64  
 7   enginesize        205 non-null    int64  
 8   boreratio         205 non-null    float64
 9   stroke            205 non-null    float64
 10  compressionratio  205 non-null    float64
 11  horsepower        205 non-null    int64  
 12  peakrpm           205 non-null    int64  
 13  citympg           205 non-null    int64  
 14  highwaympg        205 non-null    int64  
 15  price             205 non-null    int32  
 16  fuel_economy      205 non-null    float64
 1

In [26]:
cars_lr.head() 

car_ID  symboling  wheelbase  carlength  carwidth  carheight  curbweight  \
0       1          3       88.6      168.8      64.1       48.8        2548   
1       2          3       88.6      168.8      64.1       48.8        2548   
2       3          1       94.5      171.2      65.5       52.4        2823   
3       4          2       99.8      176.6      66.2       54.3        2337   
4       5          2       99.4      176.6      66.4       54.3        2824   

   enginesize  boreratio  stroke  ...  volvo  two  rear  2bbl  4bbl  idi  mfi  \
0         130       3.47    2.68  ...    0.0  1.0   0.0   0.0   0.0  0.0  0.0   
1         130       3.47    2.68  ...    0.0  1.0   0.0   0.0   0.0  0.0  0.0   
2         152       2.68    3.47  ...    0.0  1.0   0.0   0.0   0.0  0.0  0.0   
3         109       3.19    3.40  ...    0.0  0.0   0.0   0.0   0.0  0.0  0.0   
4         136       3.19    3.40  ...    0.0  0.0   0.0   0.0   0.0  0.0  0.0   

   mpfi  spdi  spfi  
0   1.0   0.0   0.0  
1   1.0   0.0   0.0  
2   1.0   0.0   0.0  
3   1.0   0.0   0.0  
4   1.0   0.0   0.0  

[5 rows x 69 columns]

In [27]:
cars_lr.shape

(205, 69)

TRAIN - TEST SPLIT AND FEATIRE SCALING

In [28]:
from sklearn.model_selection import train_test_split

np.random.seed(0) # random.seed is used to ensure that the random numbers generated are reproducible
df_train, df_test = train_test_split(cars_lr, train_size = 0.7, test_size = 0.3, random_state = 100) # random_state is used for every time you run the code, the same random numbers are generated

In [ ]:
from sklearn.preprocessing import MinMaxScaler # to scale the numerical columns to a range between 0 and 1

scaler = MinMaxScaler()
num_vars = ['wheelbase', 'carheight', 'stroke', 'curbweight', 'enginesize', 'boreratio', 'horsepower','fuel_economy','carlength','carwidth','price']
df_train[num_vars] = scaler.fit_transform(df_train[num_vars]) 

In [31]:
df_train.head() # check the scaled values in the numerical columns

car_ID  symboling  wheelbase  carlength  carwidth  carheight  curbweight  \
122     123          1   0.244828   0.426016  0.291667   0.265487    0.272692   
125     126          3   0.272414   0.452033  0.666667   0.212389    0.500388   
166     167          1   0.272414   0.448780  0.308333   0.424779    0.314973   
1         2          3   0.068966   0.450407  0.316667   0.088496    0.411171   
199     200         -1   0.610345   0.775610  0.575000   0.858407    0.647401   

     enginesize  boreratio    stroke  ...  volvo  two  rear  2bbl  4bbl  idi  \
122    0.139623   0.230159  0.525253  ...    0.0  0.0   0.0   1.0   0.0  0.0   
125    0.339623   1.000000  0.464646  ...    0.0  1.0   0.0   0.0   0.0  0.0   
166    0.139623   0.444444  0.449495  ...    0.0  1.0   0.0   0.0   0.0  0.0   
1      0.260377   0.626984  0.247475  ...    0.0  1.0   0.0   0.0   0.0  0.0   
199    0.260377   0.746032  0.484848  ...    1.0  0.0   0.0   0.0   0.0  0.0   

     mfi  mpfi  spdi  spfi  
122  0.0   0.0   0.0   0.0  
125  0.0   1.0   0.0   0.0  
166  0.0   1.0   0.0   0.0  
1    0.0   1.0   0.0   0.0  
199  0.0   1.0   0.0   0.0  

[5 rows x 69 columns]

In [32]:
#Dividing data into X and y variables
y_train = df_train.pop('price') # either you can delete the price column or pop it out, pop will remove the column and return it
X_train = df_train 

BUILDING MODEL

In [34]:
from sklearn.feature_selection import RFE
from sklearn.linear_model import LinearRegression
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import variance_inflation_factor

In [35]:
def build_model(X,y):
    X = sm.add_constant(X) # Adding the constant
    lm = sm.OLS(y,X).fit() # fitting the model
    print(lm.summary())    # model summary
    return lm
def checkVIF(X):
    vif = pd.DataFrame()
    vif['Features'] = X.columns
    vif['VIF'] = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]
    vif['VIF'] = round(vif['VIF'], 2)
    vif = vif.sort_values(by = "VIF", ascending = False)
    return(vif)

CREATING THE FIRST MODEL AND CLEAN UP FEARURES WIRH COLINEARITY

In [36]:
model1 = build_model(X_train, y_train) # building the model with the training data

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.975
Model:                            OLS   Adj. R-squared:                  0.959
Method:                 Least Squares   F-statistic:                     57.58
Date:                Thu, 29 May 2025   Prob (F-statistic):           3.34e-49
Time:                        11:33:23   Log-Likelihood:                 282.04
No. Observations:                 143   AIC:                            -446.1
Df Residuals:                      84   BIC:                            -271.3
Df Model:                          58                                         
Covariance Type:            nonrobust                                         
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
const               -0.4193      0.205  

In [40]:
vif_model1=checkVIF(X_train)
vif_model1

Features   VIF
51    plymouth   inf
27        ohcf   inf
50     peugeot   inf
48  mitsubishi   inf
45      jaguar   inf
..         ...   ...
18     hardtop  4.91
59         two  4.17
47     mercury   NaN
64         mfi   NaN
67        spfi   NaN

[68 rows x 2 columns]

In [37]:
# now we have to check the VIF values for the features and drop it
X_train.columns

Index(['car_ID', 'symboling', 'wheelbase', 'carlength', 'carwidth',
       'carheight', 'curbweight', 'enginesize', 'boreratio', 'stroke',
       'compressionratio', 'horsepower', 'peakrpm', 'citympg', 'highwaympg',
       'fuel_economy', 'gas', 'turbo', 'hardtop', 'hatchback', 'sedan',
       'wagon', 'fwd', 'rwd', 'dohcv', 'l', 'ohc', 'ohcf', 'ohcv', 'rotor',
       'five', 'four', 'six', 'three', 'twelve', 'two', 'Medium', 'Highend',
       'audi', 'bmw', 'buick', 'chevrolet', 'dodge', 'honda', 'isuzu',
       'jaguar', 'mazda', 'mercury', 'mitsubishi', 'nissan', 'peugeot',
       'plymouth', 'porsche', 'renault', 'saab', 'subaru', 'toyota',
       'volkswagen', 'volvo', 'two', 'rear', '2bbl', '4bbl', 'idi', 'mfi',
       'mpfi', 'spdi', 'spfi'],
      dtype='object')

In [38]:
# these are the features with high VIF values grater than 5
X_train1 = X_train.drop(
['car_ID', 'wheelbase', 'carlength', 'compressionratio', 'horsepower', 'citympg', 'highwaympg', 'hardtop',
 'sedan','wagon','fwd','rwd','dohcv','l','ohc','ohcf','ohcv','six','Medium',
'audi','buick','chevrolet','dodge','honda','isuzu','jaguar','mazda','nissan','porsche','renault','saab',
 'subaru','toyota','volkswagen','volvo','two','2bbl','4bbl','idi','mfi','mpfi','spdi'], axis=1)

#### Re-create the model after dropping the columns with 'P>|t|' > 0.05

In [39]:
model2 = build_model(X_train1, y_train)

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.956
Model:                            OLS   Adj. R-squared:                  0.947
Method:                 Least Squares   F-statistic:                     111.3
Date:                Thu, 29 May 2025   Prob (F-statistic):           2.01e-69
Time:                        11:52:01   Log-Likelihood:                 239.59
No. Observations:                 143   AIC:                            -431.2
Df Residuals:                     119   BIC:                            -360.1
Df Model:                          23                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const           -0.4296      0.103     -4.165   

### Feature Elimination using RFE (Recursive Feature Elimination)
Recursive feature elimination (RFE) is a feature selection method that fits a model and removes the weakest feature (or features) until the specified number of features is reached. Features are ranked by the model’s coef_ or feature_importances_ attributes, and by recursively eliminating a small number of features per loop, RFE attempts to eliminate dependencies and collinearity that may exist in the model.

In [41]:
lm = LinearRegression()
lm.fit(X_train1,y_train)

LinearRegression()

In [42]:
lm.coef_ # coefficients of the features in the model

array([ 4.58877691e-03,  3.53517341e-01, -5.22018561e-02,  3.17646711e-01,
        1.05158340e+00, -2.58582321e-01, -2.29050970e-01,  4.54766679e-05,
        4.72742553e-02, -2.67033535e-02,  5.69921931e-02, -1.97448967e-02,
        3.48014019e-01,  8.53270140e-02,  1.65089749e-01,  3.59934205e-01,
       -4.23250882e-01,  8.94202833e-02,  2.09792426e-01, -8.04911693e-16,
       -8.32366733e-02, -1.00028610e-01, -3.96298870e-02,  3.70747025e-01,
        0.00000000e+00])

In [ ]:
lm = LinearRegression()
rfe = RFE(lm,n_features_to_select=10) # Recursive Feature Elimination (RFE) is a feature selection method that fits the model and removes the weakest features until the specified number of features is reached.
rfe = rfe.fit(X_train1, y_train)

In [44]:
X_train1.columns[rfe.support_] # these are the features selected by RFE rfe.support means the features that are selected by RFE

Index(['carwidth', 'curbweight', 'enginesize', 'boreratio', 'rotor', 'three',
       'twelve', 'Highend', 'bmw', 'rear'],
      dtype='object')

#### Building model using statsmodel, for the detailed statistics

In [45]:
X_train_rfe = X_train[X_train1.columns[rfe.support_]]
X_train_rfe.head()

carwidth  curbweight  enginesize  boreratio  rotor  three  twelve  \
122  0.291667    0.272692    0.139623   0.230159    0.0    0.0     0.0   
125  0.666667    0.500388    0.339623   1.000000    0.0    0.0     0.0   
166  0.308333    0.314973    0.139623   0.444444    0.0    0.0     0.0   
1    0.316667    0.411171    0.260377   0.626984    0.0    0.0     0.0   
199  0.575000    0.647401    0.260377   0.746032    0.0    0.0     0.0   

     Highend  bmw  rear  
122      0.0  0.0   0.0  
125      1.0  0.0   0.0  
166      0.0  0.0   0.0  
1        0.0  0.0   0.0  
199      0.0  0.0   0.0

Create Model using RFE returned Features

In [46]:
model4 = build_model(X_train_rfe,y_train)

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.926
Model:                            OLS   Adj. R-squared:                  0.920
Method:                 Least Squares   F-statistic:                     164.6
Date:                Thu, 29 May 2025   Prob (F-statistic):           1.94e-69
Time:                        12:18:08   Log-Likelihood:                 202.89
No. Observations:                 143   AIC:                            -383.8
Df Residuals:                     132   BIC:                            -351.2
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.1329      0.018     -7.455      0.0

In [47]:
X_train_new = X_train_rfe.drop(["twelve"], axis = 1) # dropping the 'twelve' column as it has high p-value

Recreate model

In [48]:
model5 = build_model(X_train_new, y_train)

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.925
Model:                            OLS   Adj. R-squared:                  0.920
Method:                 Least Squares   F-statistic:                     183.2
Date:                Thu, 29 May 2025   Prob (F-statistic):           1.99e-70
Time:                        12:18:41   Log-Likelihood:                 202.50
No. Observations:                 143   AIC:                            -385.0
Df Residuals:                     133   BIC:                            -355.4
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.1323      0.018     -7.435      0.0

In [49]:
X_train_new = sm.add_constant(X_train_new) # Adding the constant
y_train_pred = model5.predict(X_train_new)

Prediction using Test Data

In [51]:
df_test

car_ID  symboling  wheelbase  carlength  carwidth  carheight  curbweight  \
160     161          0       95.7      166.3      64.4       53.0        2094   
186     187          2       97.3      171.7      65.5       55.7        2275   
59       60          1       98.8      177.8      66.5       53.7        2385   
165     166          1       94.5      168.7      64.0       52.6        2265   
140     141          2       93.3      157.3      63.8       55.7        2240   
..      ...        ...        ...        ...       ...        ...         ...   
28       29         -1      103.3      174.6      64.6       59.8        2535   
29       30          3       95.9      173.2      66.3       50.2        2811   
182     183          2       97.3      171.7      65.5       55.7        2261   
40       41          0       96.5      175.4      62.5       54.1        2372   
128     129          3       89.5      168.9      65.0       51.6        2800   

     enginesize  boreratio  stroke  ...  volvo  two  rear  2bbl  4bbl  idi  \
160          98       3.19    3.03  ...    0.0  0.0   0.0   1.0   0.0  0.0   
186         109       3.19    3.40  ...    0.0  0.0   0.0   0.0   0.0  0.0   
59          122       3.39    3.39  ...    0.0  1.0   0.0   1.0   0.0  0.0   
165          98       3.24    3.08  ...    0.0  1.0   0.0   0.0   0.0  0.0   
140         108       3.62    2.64  ...    0.0  1.0   0.0   1.0   0.0  0.0   
..          ...        ...     ...  ...    ...  ...   ...   ...   ...  ...   
28          122       3.34    3.46  ...    0.0  0.0   0.0   1.0   0.0  0.0   
29          156       3.60    3.90  ...    0.0  1.0   0.0   0.0   0.0  0.0   
182          97       3.01    3.40  ...    0.0  1.0   0.0   0.0   0.0  1.0   
40          110       3.15    3.58  ...    0.0  0.0   0.0   0.0   0.0  0.0   
128         194       3.74    2.90  ...    0.0  1.0   1.0   0.0   0.0  0.0   

     mfi  mpfi  spdi  spfi  
160  0.0   0.0   0.0   0.0  
186  0.0   1.0   0.0   0.0  
59   0.0   0.0   0.0   0.0  
165  0.0   1.0   0.0   0.0  
140  0.0   0.0   0.0   0.0  
..   ...   ...   ...   ...  
28   0.0   0.0   0.0   0.0  
29   1.0   0.0   0.0   0.0  
182  0.0   0.0   0.0   0.0  
40   0.0   0.0   0.0   0.0  
128  0.0   1.0   0.0   0.0  

[62 rows x 69 columns]

In [52]:
# Scaling the test set

num_vars = ['carwidth', 'curbweight', 'enginesize', 'boreratio', 'price', 'rotor', 'three', 'Highend', 'bmw', 'rear']

df_test1 = pd.DataFrame(scaler.fit_transform(df_test[num_vars]), columns=num_vars)

In [53]:
df_test1

carwidth  curbweight  enginesize  boreratio     price  rotor  three  \
0   0.200000    0.132148    0.082969   0.515873  0.058474    0.0    0.0   
1   0.315789    0.219125    0.131004   0.515873  0.077398    0.0    0.0   
2   0.421053    0.271985    0.187773   0.674603  0.086148    0.0    0.0   
3   0.157895    0.214320    0.082969   0.555556  0.097473    0.0    0.0   
4   0.136842    0.202307    0.126638   0.857143  0.055099    0.0    0.0   
..       ...         ...         ...        ...       ...    ...    ...   
57  0.221053    0.344065    0.187773   0.634921  0.088048    0.0    0.0   
58  0.400000    0.476694    0.336245   0.841270  0.189120    0.0    0.0   
59  0.315789    0.212398    0.078603   0.373016  0.059399    0.0    0.0   
60  0.000000    0.265738    0.135371   0.484127  0.122397    0.0    0.0   
61  0.263158    0.471408    0.502183   0.952381  0.790705    0.0    0.0   

    Highend  bmw  rear  
0       0.0  0.0   0.0  
1       0.0  0.0   0.0  
2       0.0  0.0   0.0  
3       0.0  0.0   0.0  
4       0.0  0.0   0.0  
..      ...  ...   ...  
57      0.0  0.0   0.0  
58      0.0  0.0   0.0  
59      0.0  0.0   0.0  
60      0.0  0.0   0.0  
61      1.0  0.0   1.0  

[62 rows x 10 columns]

In [54]:
#Dividing into X and y

y_test = df_test1.pop('price')
X_test = df_test1

In [55]:
y_test

0     0.058474
1     0.077398
2     0.086148
3     0.097473
4     0.055099
        ...   
57    0.088048
58    0.189120
59    0.059399
60    0.122397
61    0.790705
Name: price, Length: 62, dtype: float64

In [56]:
# Now let's use our model to make predictions.

X_test_new = pd.DataFrame(sm.add_constant(X_test))

In [57]:
X_test_new

const  carwidth  curbweight  enginesize  boreratio  rotor  three  Highend  \
0     1.0  0.200000    0.132148    0.082969   0.515873    0.0    0.0      0.0   
1     1.0  0.315789    0.219125    0.131004   0.515873    0.0    0.0      0.0   
2     1.0  0.421053    0.271985    0.187773   0.674603    0.0    0.0      0.0   
3     1.0  0.157895    0.214320    0.082969   0.555556    0.0    0.0      0.0   
4     1.0  0.136842    0.202307    0.126638   0.857143    0.0    0.0      0.0   
..    ...       ...         ...         ...        ...    ...    ...      ...   
57    1.0  0.221053    0.344065    0.187773   0.634921    0.0    0.0      0.0   
58    1.0  0.400000    0.476694    0.336245   0.841270    0.0    0.0      0.0   
59    1.0  0.315789    0.212398    0.078603   0.373016    0.0    0.0      0.0   
60    1.0  0.000000    0.265738    0.135371   0.484127    0.0    0.0      0.0   
61    1.0  0.263158    0.471408    0.502183   0.952381    0.0    0.0      1.0   

    bmw  rear  
0   0.0   0.0  
1   0.0   0.0  
2   0.0   0.0  
3   0.0   0.0  
4   0.0   0.0  
..  ...   ...  
57  0.0   0.0  
58  0.0   0.0  
59  0.0   0.0  
60  0.0   0.0  
61  0.0   1.0  

[62 rows x 10 columns]

In [58]:
# Making predictions
y_pred = model5.predict(X_test_new)

Model Performance Comparisn between Train and Test (Generalisation)

In [59]:
from sklearn.metrics import r2_score
print("Test Prediction R-Sqrd: ", r2_score(y_test, y_pred))

Test Prediction R-Sqrd:  0.8606328550795905


In [60]:
print("Train Prediction R-Sqrd: ", r2_score(y_train, y_train_pred))

Train Prediction R-Sqrd:  0.9253685182104617


#### Inference :

- R-sqaured value for Training data is `92.5%` and for Test Data it is `86%`.
- This scores are fairly decent, however, you can investigate further to achieve better generalisation.

### OUR FINAL MODEL

In [61]:
print(model5.summary())

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.925
Model:                            OLS   Adj. R-squared:                  0.920
Method:                 Least Squares   F-statistic:                     183.2
Date:                Thu, 29 May 2025   Prob (F-statistic):           1.99e-70
Time:                        12:27:00   Log-Likelihood:                 202.50
No. Observations:                 143   AIC:                            -385.0
Df Residuals:                     133   BIC:                            -355.4
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.1323      0.018     -7.435      0.0